cardano-crypto-praos/src/Cardano/Crypto/VRF/Praos.hs

```diff
@@ -0,0 +1,454 @@
+    return (pk, sk)
+
+-- | Derive a Public Key from a Secret Key.
+skToPK :: SK -> PK
+skToPK sk =
+  unsafePerformIO $ withForeignPtr (unSK sk) $ \skPtr -> do
+    pk <- mkPK
+    withForeignPtr (unPK pk) $ \pkPtr -> do
+      void $ crypto_vrf_sk_to_pk pkPtr skPtr
+    return pk
+
+-- | Get the seed used to generate a given Secret Key
+skToSeed :: SK -> Seed
+skToSeed sk =
+  unsafePerformIO $ withForeignPtr (unSK sk) $ \skPtr -> do
+    seed <- mkSeed
+    _ <- withForeignPtr (unSeed seed) $ \seedPtr -> do
+      crypto_vrf_sk_to_seed seedPtr skPtr
+    return seed
+
+-- | Construct a proof from a Secret Key and a message.
+-- Returns 'Just' the proof on success, 'Nothing' if the secrect key could not
+-- be decoded.
+prove :: SK -> ByteString -> Maybe Proof
+prove sk msg =
+  unsafePerformIO $
+    withForeignPtr (unSK sk) $ \skPtr -> do
+      proof <- mkProof
+      BS.useAsCStringLen msg $ \(m, mlen) -> do
+        withForeignPtr (unProof proof) $ \proofPtr -> do
+          crypto_vrf_prove proofPtr skPtr m (fromIntegral mlen) >>= \case
+            0 -> return $ Just proof
+            _ -> return Nothing
+
+-- | Verify a VRF proof and validate the Public Key. Returns 'Just' a hash of
+-- the verification result on success, 'Nothing' if the verification did not
+-- succeed.
+--
+-- For a given public key and message, there are many possible proofs but only
+-- one possible output hash.
+verify :: PK -> Proof -> ByteString -> Maybe Output
+verify pk proof msg =
+  unsafePerformIO $
+    withForeignPtr (unPK pk) $ \pkPtr -> do
+      withForeignPtr (unProof proof) $ \proofPtr -> do
+        output <- mkOutput
+        BS.useAsCStringLen msg $ \(m, mlen) -> do
+          withForeignPtr (unOutput output) $ \outputPtr -> do
+            crypto_vrf_verify outputPtr pkPtr proofPtr m (fromIntegral mlen) >>= \case
+              0 -> return $ Just output
+              _ -> return Nothing
+
+data PraosVRF
+
+instance VRFAlgorithm PraosVRF where
+  newtype VerKeyVRF PraosVRF = VerKeyPraosVRF PK
+    deriving stock   (Show, Eq, Generic)
+    deriving newtype (ToCBOR, FromCBOR)
+    deriving NoUnexpectedThunks via UseIsNormalForm (ForeignPtr PKValue)
+
+  newtype SignKeyVRF PraosVRF = SignKeyPraosVRF SK
+    deriving stock   (Show, Eq, Generic)
+    deriving newtype (ToCBOR, FromCBOR)
+    deriving NoUnexpectedThunks via UseIsNormalForm (ForeignPtr SKValue)
+
+  newtype CertVRF PraosVRF = CertPraosVRF Proof
+    deriving stock   (Show, Eq, Generic)
+    deriving newtype (ToCBOR, FromCBOR)
+    deriving NoUnexpectedThunks via UseIsNormalForm (ForeignPtr ProofValue)
+
+  type Signable PraosVRF = ToCBOR
+
+  algorithmNameVRF = const "PraosVRF"
+
+  deriveVerKeyVRF = coerce skToPK
+
+  evalVRF = \_ msg (SignKeyPraosVRF sk) -> do
+    let msgBS = serialize' msg
+    proof <- maybe (error "Invalid Key") pure $ prove sk msgBS
+    output <- maybe (error "Invalid Proof") pure $ verify (skToPK sk) proof msgBS
```

Comment:
We're doing more work than we need here, running an extra verify round just to extract the hash. Instead, we should just call `crypto_vrf_proof_to_hash` to get the output (https://github.com/input-output-hk/libsodium/blob/tdammers/rebased-vrf/src/libsodium/crypto_vrf/crypto_vrf.c#L75).

Reply:
Done.